dali/test/python/test_operator_noise_salt_and_pepper.py

```diff
@@ -0,0 +1,85 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nvidia.dali import pipeline_def
+from nvidia.dali.backend_impl import TensorListGPU
+import nvidia.dali.ops as ops
+import nvidia.dali.fn as fn
+import nvidia.dali.types as types
+import random
+import numpy as np
+import os
+from test_utils import get_dali_extra_path, check_batch
+from test_noise_utils import PSNR
+
+test_data_root = get_dali_extra_path()
+images_dir = os.path.join(test_data_root, 'db', 'single', 'png')
+dump_images = False
+
+def salt_and_pepper_noise_ref(x, prob, salt_to_pepper_prob):
+    x = np.array(x, dtype=np.float32)
+    salt_prob = prob * salt_to_pepper_prob
+    pepper_prob = prob * (1.0 - salt_to_pepper_prob)
+    mask = np.random.choice(
+        [0., 1., np.nan], p=[pepper_prob, 1 - prob, salt_prob], size=x.shape
+    )
+    y = np.where(np.isnan(mask), 255., x * mask).astype(np.uint8)
+    return y
+
+@pipeline_def
+def pipe_salt_and_pepper_noise(prob, salt_to_pepper_prob, device='cpu'):
+    encoded, _ = fn.readers.file(file_root=images_dir)
+    in_data = fn.decoders.image(encoded, device="cpu", output_type=types.RGB)
+    if device == 'gpu':
+        in_data = in_data.gpu()
+    prob_arg = prob or fn.random.uniform(range=(0.05, 0.5))
+    salt_to_pepper_prob_arg = salt_to_pepper_prob or fn.random.uniform(range=(0.25, 0.75))
+    out_data = fn.noise.salt_and_pepper(
+        in_data, prob=prob_arg, salt_to_pepper_prob=salt_to_pepper_prob_arg)
+    return in_data, out_data, prob_arg, salt_to_pepper_prob_arg
+
+def _testimpl_operator_noise_salt_and_pepper(device, prob, salt_to_pepper_prob, batch_size, niter):
+    pipe = pipe_salt_and_pepper_noise(prob, salt_to_pepper_prob, device=device, batch_size=batch_size,
+                                      num_threads=3, device_id=0, seed=12345)
+    pipe.build()
+    for _ in range(niter):
+        out_data, in_data, prob_arg, salt_to_pepper_prob_arg = pipe.run()
+        prob_arg = prob_arg.as_array()
+        salt_to_pepper_prob_arg = salt_to_pepper_prob_arg.as_array()
+        if device == 'gpu':
+            out_data = out_data.as_cpu()
+            in_data = in_data.as_cpu()
+        for s in range(batch_size):
+            sample_in = np.array(out_data[s])
+            sample_out = np.array(in_data[s])
+            prob = float(prob_arg[s])
+            salt_to_pepper_prob = float(salt_to_pepper_prob_arg[s])
+            sample_ref = salt_and_pepper_noise_ref(sample_in, prob, salt_to_pepper_prob)
+            psnr_out = PSNR(sample_out, sample_in)
+            psnr_ref = PSNR(sample_ref, sample_in)
```

Comment:
You can also check that where the output doesn't match the reference you either get salt or pepper, and not any other value (maybe np.where(sample_out!+sample_in) could help).
You can also check the distribution of salt and pepper - should be pretty easy.

Reply:
done